File: GYM Questions/Hashing/Find pairs with given sum.java
// https://course.acciojob.com/idle?question=956f9a39-b446-4253-8df9-ff673dadc1ed

import java.io.*;
import java.util.*;
public class Main {
    
    public static int getPairsCount(int[] arr, int n, int sum){
		int count  = 0;
	    HashMap<Integer, Integer> storage = new HashMap<>();

		for(int num : arr){
			if(storage.containsKey(sum-num))
				count += storage.get(sum-num);

			storage.put(num, storage.getOrDefault(num,0)+1);
		}
		return count;
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int sum = sc.nextInt();
        int arr[] = new int[n];
        for(int i=0; i<n; i++)
        {
            arr[i]=sc.nextInt();
        }
        System.out.println(getPairsCount(arr,n, sum));
    }
}

File: Javascript/Find pairs with given sum.js
// https://course.acciojob.com/idle?question=c3287644-0a7d-407f-9a57-bb7899ca62ff

function countPairs(n, sum, arr) {
 // return the count
// do not console.log here
    const storage = new Map();
    let count = 0;

    for(let num of arr){
        if(storage.has(sum-num))
            count += storage.get(sum-num);

        if(storage.has(num)) storage.set(num, storage.get(num)+1); 
        else storage.set(num,1);
    }
    return count;
}

/*Do not change the code below */

var readline = require("readline").createInterface(process.stdin);

let inputArr = [];
var lineNumber = -1;
var inputSize;

//multipleline input from User
readline.on("line", readInputs);

function readInputs(line) {
  inputArr.push(line);
  lineNumber++;
  if (lineNumber == 0) {
    size = parseInt(inputArr[0]);
  }
  //Exit Condition
  if (lineNumber == 1) {
    logic("s");
    readline.close();
  }
}

function logic(input) {
  //let T = parseInt(inputArr[0].trim());
  //let str1 = inputArr[1].trim();
  let [n, k] = inputArr[0].split(" ").map((x) => parseInt(x));
  let Arr = inputArr[1]
    .trim()
    .split(" ")
    .map((x) => parseInt(x));
  console.log(countPairs(n, k, Arr));
}
